packages/contracts-bedrock/src/L1/OptimismPortal2.sol

```diff
@@ -162,74 +152,143 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ISemver {
+    /// @notice Thrown when a withdrawal has not been proven for long enough.
+    error OptimismPortal_ProofNotOldEnough();
+
+    /// @notice Thrown when a withdrawal has not been proven.
+    error OptimismPortal_Unproven();
+
+    /// @notice Thrown when the caller is not authorized to call the function.
+    error OptimismPortal_Unauthorized();
+
+    /// @notice Thrown when the wrong proof method is used.
+    error OptimismPortal_WrongProofMethod();
+
+    /// @notice Thrown when a super root proof is invalid.
+    error OptimismPortal_InvalidSuperRootProof();
+
+    /// @notice Thrown when an output root index is invalid.
+    error OptimismPortal_InvalidOutputRootIndex();
+
+    /// @notice Thrown when an output root chain id is invalid.
+    error OptimismPortal_InvalidOutputRootChainId();
 
     /// @notice Reverts when paused.
     modifier whenNotPaused() {
-        if (paused()) revert CallPaused();
+        if (paused()) revert OptimismPortal_CallPaused();
         _;
     }
 
     /// @notice Semantic version.
-    /// @custom:semver 3.14.0
+    /// @custom:semver 4.0.0
     function version() public pure virtual returns (string memory) {
-        return "3.14.0";
+        return "4.0.0";
     }
 
-    /// @notice Constructs the OptimismPortal contract.
-    constructor(uint256 _proofMaturityDelaySeconds, uint256 _disputeGameFinalityDelaySeconds) {
+    /// @param _proofMaturityDelaySeconds The proof maturity delay in seconds.
+    constructor(uint256 _proofMaturityDelaySeconds) ReinitializableBase(2) {
         PROOF_MATURITY_DELAY_SECONDS = _proofMaturityDelaySeconds;
-        DISPUTE_GAME_FINALITY_DELAY_SECONDS = _disputeGameFinalityDelaySeconds;
-
         _disableInitializers();
     }
 
     /// @notice Initializer.
-    /// @param _disputeGameFactory Contract of the DisputeGameFactory.
-    /// @param _systemConfig Contract of the SystemConfig.
-    /// @param _superchainConfig Contract of the SuperchainConfig.
+    /// @param _systemConfig Address of the SystemConfig.
+    /// @param _superchainConfig Address of the SuperchainConfig.
+    /// @param _anchorStateRegistry Address of the AnchorStateRegistry.
+    /// @param _ethLockbox Contract of the ETHLockbox.
+    /// @param _superRootsActive Whether the OptimismPortal is using Super Roots or Output Roots.
     function initialize(
-        IDisputeGameFactory _disputeGameFactory,
         ISystemConfig _systemConfig,
         ISuperchainConfig _superchainConfig,
-        GameType _initialRespectedGameType
+        IAnchorStateRegistry _anchorStateRegistry,
+        IETHLockbox _ethLockbox,
+        bool _superRootsActive
     )
         external
-        initializer
+        reinitializer(initVersion())
     {
-        disputeGameFactory = _disputeGameFactory;
         systemConfig = _systemConfig;
         superchainConfig = _superchainConfig;
+        anchorStateRegistry = _anchorStateRegistry;
+        ethLockbox = _ethLockbox;
```

Comment:
The system config `Addresses` struct should have the ASR and lockbox added to it, then here we can just fetch those addresses from the system config

Reply:
I agree that there should be a single source of truth to easily get all the addresses, but I'm wary of hardcoding them here as it creates one more thing we need to keep in sync.

ie. with the ASR, this upgrade is the second time we're replacing it. In such cases the upgrade would also need to modify the SystemConfig. 

With the Lockbox, we have an `updateLockbox()` method on the Portal which would become insufficient for what it does. 

What I'd rather see is a `SystemConfig.getSystem()` method which crawls the chain to return all the addresses. I recently made a [script](https://gist.github.com/maurelian/a9b7b2ccb0cca1806bdf05442038a9b1) that uses `cast` to do that, can easily be converted to a solidity library.

Reply:
Could we do this as part of a unifying upgrade that has all contracts take everything from `SystemConfig`? Changing this now will be a pain

Reply:
Am fine with that.